[SPARK-22826][SQL] findWiderTypeForTwo Fails over StructField of Array

bdrillard opened the pull request:
## What changes were proposed in this pull request?

[SPARK-22826](https://issues.apache.org/jira/browse/SPARK-22826)

Attempting to find the tightest common type over a struct holding fields of `ArrayType` or `MapType` will fail if the types of those struct fields are not exactly equal. See the JIRA for an example.

## How was this patch tested?

This patch adds a test case for `ArrayType` and `MapType` for finding tightest common types.

Commit:
[SPARK-22826] adding support for tighest type coercion over array/map types in a structfield

[SPARK-22826] adding nested struct tests

[SPARK-22826] removing extraneous test case

AmplabJenkins commented:
Can one of the admins verify this patch?

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -99,6 +99,17 @@ object TypeCoercion {
       Some(TimestampType)
 
+    case (t1 @ ArrayType(pointType1, nullable1), t2 @ ArrayType(pointType2, nullable2))
+        if t1.sameType(t2) =>
```
why is this `if t1.sameType(t2)` needed?

gczsjdy replied:
This will make sure the later `get` won't be applied to `None`

gczsjdy commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -99,6 +99,17 @@ object TypeCoercion {
 
+    case (t1 @ ArrayType(pointType1, nullable1), t2 @ ArrayType(pointType2, nullable2))
+        if t1.sameType(t2) =>
+      val dataType = findTightestCommonType(pointType1, pointType2).get
```
If `t1` and `t2` are the sameType, why do we need recursively `findTightestCommonType`? Can we just make `val dataType = pointType1`?

mgaido91 replied:
I think that the reason is for nested structures, for instance an array of array or array of map

mgaido91 replied:
then @bdrillard can we add also some test cases for nested structures?

gczsjdy replied:
Let's use this example, if we call `findTightestCommonType` on 2 array of array of Integer types, then this `dataType` = `findTightest(findTightest(IntegralType, IntegralType))` = `IntegralType`, so the final answer is `Some(ArrayType(IntegralType))`?

bdrillard replied:
@mgaido91, @gczsjdy, sure, I'll add a pair of test cases for nested complex structures for `ArrayType` and `MapType`

gczsjdy replied:
@mgaido91 Thanks, I got it.

mgaido91 replied:
@gczsjdy  I think it is needed, because IIUC in `sameType` the nullability is not checked to be the same. So it might vary and calling `findTightestCommonType` does the trick.

bdrillard replied:
@gczsjdy `sameType` will ignore the nullability of the two types, only checking if the `DataType` is the same. We would expect then that the first case of `findTightestCommonType` would pass the types on through, but during its equality check, it also checks the type nullability, which for ArrayType and MapType, may differ between t1 and t2:

```
case (t1, t2) if t1 == t1 => Some(t1)
```

That means we can establish that two StructFields of ArrayType/MapType are the same DataType, but if they have different nullability, then the above case match won't match them, nor will any other case in the match set. So in order to find the tightest common type where nullabilities of the point-types may differ, we'll need to recurse. See a case like:

```
widenTest(
    StructType(StructField("a", ArrayType(StringType, containsNull=true)) :: Nil),
    StructType(StructField("a", ArrayType(StringType, containsNull=false)) :: Nil),
    Some(StructType(StructField("a", ArrayType(StringType, containsNull=true)) :: Nil)))
```

At the moment, since we have no case to match ArrayType/MapType where the nullabilities may differ, this case would fail. I can add this test explicitly to the suite.

gczsjdy replied:
Thanks for your illustration.

gczsjdy commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -99,6 +99,17 @@ object TypeCoercion {
+    case (t1 @ MapType(keyType1, valueType1, nullable1),
+        t2 @ MapType(keyType2, valueType2, nullable2)) if t1.sameType(t2) =>
+      val keyType = findTightestCommonType(keyType1, keyType2).get
+      val valueType = findTightestCommonType(valueType1, valueType2).get
```
Also here

bdrillard commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -158,11 +169,6 @@ object TypeCoercion {
-        case (ArrayType(et1, containsNull1), ArrayType(et2, containsNull2)) =>
-          findWiderTypeForTwo(et1, et2).map(ArrayType(_, containsNull1 || containsNull2))
-        case _ => None
-      })
```
I think if we now check for ArrayTypes (including MapTypes) in `findTightestCommonType`, the match here becomes redundant. @mgaido91, @gczsjdy, does this thinking make sense to you both?

mgaido91 replied:
it makes sense, but I'd love that in your implementation in `findTightestCommonType`, you would replicate this logic, ie. removing the `sameType` guard and using `findWiderTypeForTwo`, in order to allow casting an array of int to an array of long. What do you think?

bdrillard replied:
Sure, I think that's possible. In order to handle instances with and without string promotion, I think it may be necessary to add a boolean parameter, and then to handle the instances where the pointType/keyType and valueType may result in `None`, see https://github.com/apache/spark/pull/20010/files#diff-383a8cdd0a9c58cae68e0a79295520a3R105

To support the minor change in function signature for `findTightestCommonType`, I have to do some refactoring in the tests. Let me know if you think there's a cleaner implementation, but this seems to help localize like concerns into `findTightestCommonType`.

gczsjdy replied:
Thanks for your illustration.

gczsjdy replied:
My suggestion: we define a new function, like `findWiderTypeForArray`. This new function can provide 'findWider' functionality compared to the 'findTightestArray' part(which is basically the first commit in your PR). We won't break the original `findTightest` semantic in this way and the code is clean.

mgaido91 replied:
I like your idea @gczsjdy!

bdrillard replied:
@gczsjdy I've taken a shot at implementing your suggestion with `findWiderTypeForTwoComplex`, which takes as an argument a `widerTypeFunc`, describing which widening behavior to apply to point types (should they permit promotion to string or not). 

Because `ArrayType` instances that would require widening the type could be nested in `StructType` and `MapType`, I think it's necessary to have more case matching than would be in `findWiderTypeForArray`, hence `findWiderTypeForTwoComplex`.

gczsjdy replied:
I like your implementation, except for the concerns I raised in another thread. Have you tested locally?

bdrillard replied:
Yes, this PR should be ready for a Jenkins build.

bdrillard commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -182,12 +188,6 @@ object TypeCoercion {
-          findWiderTypeWithoutStringPromotionForTwo(et1, et2)
-            .map(ArrayType(_, containsNull1 || containsNull2))
-        case _ => None
-      })
```
Same comment here, as above.

bdrillard commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala:
```diff
@@ -389,6 +389,25 @@ class TypeCoercionSuite extends AnalysisTest {
+      StructType(StructField("a", MapType(StringType, StringType, valueContainsNull=true)) :: Nil),
+      StructType(StructField("a", MapType(StringType, StringType, valueContainsNull=false)) :: Nil),
+      Some(StructType(
+        StructField("a", MapType(StringType, StringType, valueContainsNull=true)) :: Nil)))
```
Here's a test for nested structures where an explicit match case against ArrayType/MapType is necessary due to the difference in nullability between the two structures.

mgaido91 replied:
thanks, may you please add also two tests having Map and Array as outer types?

Commit:
[SPARK-22826] supporting with and without string promition

[SPARK-22826] adding test for integral type promotion

[SPARK-22826] supporting array and map string promotion

gczsjdy commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -99,6 +102,33 @@ object TypeCoercion {
 
+    case (ArrayType(pointType1, nullable1), ArrayType(pointType2, nullable2)) =>
+      val dataType = if (withStringPromotion) {
+        findWiderTypeForTwo(pointType1, pointType2)
```
I think we break the `findTightest` semantic by `withStringPromotion` judgement and calling `findWiderTypeForTwo`, these are what we should add in `Case 2 type widening` -> `findWiderTypeForTwo`. I suggest we put these logic in another function`findWiderTypeForDecimal`. Will mention this in the other thread.

Commit:
Revert "[SPARK-22826] supporting with and without string promition"

This reverts commit da36de8b4035e143efa388b277596dceb185130a.

Commit:
[SPARK-22826] cleaning up api over wider types

gczsjdy commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -158,11 +213,8 @@ object TypeCoercion {
-          findWiderTypeForTwo(et1, et2).map(ArrayType(_, containsNull1 || containsNull2))
-        case _ => None
-      })
+      .orElse(findWiderTypeForTwoComplex(t1, t2, findWiderTypeForTwo))
```
Will this cause infinite loop? Calling `findWiderTypeForTwo` in `findWiderTypeForTwo`?

bdrillard commented:
@gczsjdy, @mgaido91 If you all are comfortable with it, I think this PR is in a state where we could trigger a build.

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -158,11 +213,8 @@ object TypeCoercion {
-        case _ => None
-      })
+      .orElse(findWiderTypeForTwoComplex(t1, t2, findWiderTypeForTwo))
+
```
this line should be removed

bdrillard replied:
Fixed, see dfb4a3b47d88853be348aafd9802f799639693f6

mgaido91 commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala:
```diff
@@ -389,6 +389,25 @@ class TypeCoercionSuite extends AnalysisTest {
     widenTest(ArrayType(IntegerType), StructType(Seq()), None)
 
+    widenTest(
+      ArrayType(StringType, containsNull=true),
```
please add a whitespace before and after `=` and the same in the following lines

bdrillard replied:
Fixed, see dfb4a3b47d88853be348aafd9802f799639693f6

mgaido91 commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala:
```diff
@@ -389,6 +389,25 @@ class TypeCoercionSuite extends AnalysisTest {
+    widenTest(
+      MapType(StringType, StringType, valueContainsNull=true),
+      MapType(StringType, StringType, valueContainsNull=false),
+      Some(MapType(StringType, StringType, valueContainsNull=true)))
```
please add a test also for struct

bdrillard replied:
Here are tests for nested complex types, https://github.com/apache/spark/pull/20010/files#diff-01ecdd038c5c2f53f38118912210fef8R405

mgaido91 commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala:
```diff
@@ -469,12 +488,21 @@ class TypeCoercionSuite extends AnalysisTest {
+    widenTestWithStringPromotion(
+      StructType(StructField("a", ArrayType(LongType)) :: Nil),
+      StructType(StructField("a", ArrayType(StringType)) :: Nil),
+      Some(StructType(StructField("a", ArrayType(StringType)) :: Nil)))
```
please add an analogous test also for map and array

bdrillard replied:
I've grouped string promotion tests for arrays and maps, with tests where those types are nested in structs or are fields in structs, https://github.com/apache/spark/pull/20010/files#diff-01ecdd038c5c2f53f38118912210fef8R504

mgaido91 commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala:
```diff
@@ -469,12 +488,21 @@ class TypeCoercionSuite extends AnalysisTest {
+    widenTestWithoutStringPromotion(
+      StructType(StructField("a", ArrayType(LongType)) :: Nil),
+      StructType(StructField("a", ArrayType(StringType)) :: Nil),
+      None)
```
ditto

bdrillard replied:
I've grouped non-string promotion tests, and added tests for arrays, maps, and then structures nested in those types or having fields of those types, https://github.com/apache/spark/pull/20010/files#diff-01ecdd038c5c2f53f38118912210fef8R563

mgaido91 commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala:
```diff
@@ -469,12 +488,21 @@ class TypeCoercionSuite extends AnalysisTest {
+      StructType(StructField("a", ArrayType(LongType)) :: Nil),
+      StructType(StructField("a", ArrayType(StringType)) :: Nil),
+      Some(StructType(StructField("a", ArrayType(StringType)) :: Nil)))
+
```
this empty line is not needed

bdrillard replied:
Fixed, see dfb4a3b47d88853be348aafd9802f799639693f6

mgaido91 commented:
@bdrillard yes, after addressing my last comments I think we can build this. My only concern is that we are not covering the case in which there are nested complex structures in the wider case. But maybe we can address this later.

I don't have permissions to trigger a build. You might check who are the last committers working on this and maybe ask them.
Thanks.

Commit:
[SPARK-22826] adding more test cases

bdrillard commented:
@mgaido91 Agreed with that concern. I think the last round of tests I've just added covers the permutation of cases well, where we have arrays and maps of structs, and structs of arrays and maps.

cc: @viirya, @gatorsmile, Could perhaps one of you trigger a Jenkins build against this PR?

viirya commented:
@HyukjinKwon Can you help trigger Jenkins test for this? Thanks.

HyukjinKwon commented:
ok to test

SparkQA commented:
**[Test build #85447 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85447/testReport)** for PR 20010 at commit [`8ce3b39`](https://github.com/apache/spark/commit/8ce3b39369266b222323561afd431ecfe0dc5270).

SparkQA commented:
**[Test build #85447 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85447/testReport)** for PR 20010 at commit [`8ce3b39`](https://github.com/apache/spark/commit/8ce3b39369266b222323561afd431ecfe0dc5270).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

viirya commented:
Thanks @HyukjinKwon

gczsjdy commented:
Seems not a regular error? 
@bdrillard Maybe you can push a commit and trigger the test again.

SparkQA commented:
**[Test build #85497 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85497/testReport)** for PR 20010 at commit [`86e1929`](https://github.com/apache/spark/commit/86e1929c490861d9e93ef34abd52c442f99f31a9).

SparkQA commented:
**[Test build #85497 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85497/testReport)** for PR 20010 at commit [`86e1929`](https://github.com/apache/spark/commit/86e1929c490861d9e93ef34abd52c442f99f31a9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85497/
Test FAILed.

mgaido91 commented:
@gczsjdy @bdrillard the test errors are valid. In some cases an exception is expected to be thrown, but it isn't, due to the fix. So they should be updated accordingly.

Commit:
[SPARK-22826] checking to sameType on complex types

SparkQA commented:
**[Test build #85514 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85514/testReport)** for PR 20010 at commit [`583f674`](https://github.com/apache/spark/commit/583f6748e5678417fce817b829acf6151f166e92).

SparkQA commented:
**[Test build #85514 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85514/testReport)** for PR 20010 at commit [`583f674`](https://github.com/apache/spark/commit/583f6748e5678417fce817b829acf6151f166e92).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85514/
Test FAILed.

Commit:
[SPARK-22826] fixup on spacing

SparkQA commented:
**[Test build #85515 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85515/testReport)** for PR 20010 at commit [`23c83ac`](https://github.com/apache/spark/commit/23c83ac7e13bbf23347ff14b65a2413342aae00f).

Commit:
[SPARK-22826] reverting use of sameType

Commit:
[SPARK-22826] adding sensitivity to field names and case

SparkQA commented:
**[Test build #85520 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85520/testReport)** for PR 20010 at commit [`e034bbc`](https://github.com/apache/spark/commit/e034bbcad245e57c8c0276b669303c1d70504f5a).

SparkQA commented:
**[Test build #85520 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85520/testReport)** for PR 20010 at commit [`e034bbc`](https://github.com/apache/spark/commit/e034bbcad245e57c8c0276b669303c1d70504f5a).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85520/
Test FAILed.

Commit:
[SPARK-22826] style fixup

SparkQA commented:
**[Test build #85521 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85521/testReport)** for PR 20010 at commit [`7d146e3`](https://github.com/apache/spark/commit/7d146e313cfd9ecf8c47f046028d9ff70f0301d3).

SparkQA commented:
**[Test build #85515 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85515/testReport)** for PR 20010 at commit [`23c83ac`](https://github.com/apache/spark/commit/23c83ac7e13bbf23347ff14b65a2413342aae00f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85515/
Test FAILed.

SparkQA commented:
**[Test build #85521 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85521/testReport)** for PR 20010 at commit [`7d146e3`](https://github.com/apache/spark/commit/7d146e313cfd9ecf8c47f046028d9ff70f0301d3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85521/
Test FAILed.

Commit:
[SPARK-22826] fixing issue with boolean condition on case sensitivity

SparkQA commented:
**[Test build #85526 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85526/testReport)** for PR 20010 at commit [`d42dfa5`](https://github.com/apache/spark/commit/d42dfa55105c8944b63781fc61b59c98a99338d1).

SparkQA commented:
**[Test build #85526 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85526/testReport)** for PR 20010 at commit [`d42dfa5`](https://github.com/apache/spark/commit/d42dfa55105c8944b63781fc61b59c98a99338d1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85526/
Test PASSed.

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -148,6 +160,61 @@ object TypeCoercion {
+        }
+
+      case (StructType(fields1), StructType(fields2)) =>
+        val fieldTypes = fields1.zip(fields2).map { case (f1, f2) =>
```
this requires that fields with the same name must also be in the same position. Is this assumption correct?

bdrillard replied:
@mgaido91 That's seems to be the assumption already made in [`findTightestCommonType`](https://github.com/bdrillard/spark/blob/d42dfa55105c8944b63781fc61b59c98a99338d1/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L115). The [`sameType`](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala#L83) function on DataType also requires structfields are ordered the same. 

The difference here is that we don't require the structfields strictly have the same type, so we can support widening to LongType, StringType, etc. But we do require the fields 1. have the same order, and 2. have the same name (either with strict case, or ignoring case).

marmbrus commented:
/cc @cloud-fan @sameeragarwal

gatorsmile commented:
Will review it today.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -99,11 +100,22 @@ object TypeCoercion {
     case (_: TimestampType, _: DateType) | (_: DateType, _: TimestampType) =>
       Some(TimestampType)
 
+    case (t1 @ ArrayType(pointType1, nullable1), t2 @ ArrayType(pointType2, nullable2))
```
`pointType` looks like a weird name, how about `elementType` or `et`?

bdrillard replied:
Sure, see #09e49fb8162b97dec76a6324fa4fac4553b22013

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala:
```diff
@@ -462,27 +507,139 @@ class TypeCoercionSuite extends AnalysisTest {
-      ArrayType(StringType, containsNull = true),
+      ArrayType(ArrayType(IntegerType), containsNull = true),
+      ArrayType(ArrayType(LongType), containsNull = false),
+      Some(ArrayType(ArrayType(LongType), containsNull = true)))
```
Do we really wanna do this? Is there any other database have this behavior?

I think for complex type, we should ignore the nullable difference, but I'm not sure if we should do type coercion inside complex type.

bdrillard replied:
@mgaido91, thoughts on this?

It's definitely possible for us to revert back to behavior where we don't do IntegerType-to-LongType, xType-to-StringType, etc. promotion inside complex types, which was how a previous form of this PR handled it.

mgaido91 replied:
I think this is more coherent with the behavior in other parts and this is the behavior I would expect. But I think that we should follow @gatorsmile's suggestion and check Hive's behavior first.

Commit:
[SPARK-22826] changing array type element name

SparkQA commented:
**[Test build #85682 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85682/testReport)** for PR 20010 at commit [`09e49fb`](https://github.com/apache/spark/commit/09e49fb8162b97dec76a6324fa4fac4553b22013).

SparkQA commented:
**[Test build #85682 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85682/testReport)** for PR 20010 at commit [`09e49fb`](https://github.com/apache/spark/commit/09e49fb8162b97dec76a6324fa4fac4553b22013).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85682/
Test PASSed.

gatorsmile commented:
Overall, it is reasonable. What is the current behavior in Hive?

bdrillard commented:
Is there a consensus on the preferred behavior here? This issue would also be a blocker to encoders for Spark-Avro in Spark 2.3.0 that @marmbrus mentions in #20085.

mgaido91 commented:
@bdrillard I think that we need to know Hive's behavior as per https://github.com/apache/spark/pull/20010#issuecomment-355717067 to state which is the preferred/right behavior here.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
ok to test

SparkQA commented:
**[Test build #91613 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91613/testReport)** for PR 20010 at commit [`09e49fb`](https://github.com/apache/spark/commit/09e49fb8162b97dec76a6324fa4fac4553b22013).

SparkQA commented:
**[Test build #91613 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91613/testReport)** for PR 20010 at commit [`09e49fb`](https://github.com/apache/spark/commit/09e49fb8162b97dec76a6324fa4fac4553b22013).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91613/
Test FAILed.

AmplabJenkins commented:
Can one of the admins verify this patch?
